src/nxt_otel.c

```diff
@@ -0,0 +1,273 @@
+    {
+        nxt_log(t, NXT_LOG_ERR, "no trace to add events to!");
+        nxt_otel_state_transition(r->otel, NXT_OTEL_ERROR_STATE);
+        return;
+    }
+
+    nxt_list_each(cur, r->fields) {
+        // we need this in a continguous and null terminated segment of memory for Rust FFI
+        name_cur = nxt_mp_zalloc(r->mem_pool, cur->name_length + 1);
+        val_cur = nxt_mp_zalloc(r->mem_pool, cur->value_length + 1);
+        if (name_cur && val_cur) {
+            strncpy((char *) name_cur, (char *) cur->name, cur->name_length);
+            strncpy((char *) val_cur, (char *) cur->value, cur->value_length);
+            nxt_otel_add_event_to_trace(r->otel->trace, name_cur, val_cur);
+        }
+    } nxt_list_loop;
+
+    traceval = nxt_mp_zalloc(r->mem_pool, NXT_OTEL_TRACEPARENT_LEN + 1);
+    if (!traceval) {
+        /* let it go blank here.
+         * span still gets populated and sent
+         * but data is not propagated to peer or app.
+         */
+        nxt_log(t, NXT_LOG_ERR,
+                "couldnt allocate traceparent header. span will not propagate");
+        return;
+    }
+
+    // if we didnt inherit a trace id then we need to add the
+    // traceparent header to the request
+    if (!r->otel->trace_id) {
+        nxt_otel_copy_traceparent(traceval, r->otel->trace);
+        f = nxt_list_add(r->fields);
+        if (f) {
+            nxt_http_field_name_set(f, "traceparent");
+            f->value = traceval;
+            f->value_length = nxt_strlen(traceval);
+            nxt_otel_add_event_to_trace(r->otel->trace, f->name, traceval);
+        } else {
+            nxt_log(t, NXT_LOG_ERR,
+                    "couldnt allocate traceparent header in request");
+        }
+    } else {
+        // copy in the pre-existing traceparent for the response
+        snprintf((char *) traceval, NXT_OTEL_TRACEPARENT_LEN + 1, "%s-%s-%s-%s",
+                 (char *) r->otel->version,
+                 (char *) r->otel->trace_id,
+                 (char *) r->otel->parent_id,
+                 (char *) r->otel->trace_flags);
+    }
+
+    f = NULL;
+    f = nxt_list_add(r->resp.fields);
+    if (f) {
+        nxt_http_field_name_set(f, "traceparent");
+        f->value = traceval;
+        f->value_length = nxt_strlen(traceval);
+    } else {
+        nxt_log(t, NXT_LOG_ERR,
+                "couldnt allocate traceparent header in response");
+    }
+
+    nxt_otel_state_transition(r->otel, NXT_OTEL_BODY_STATE);
+}
+
+static void
+nxt_otel_span_add_body(nxt_http_request_t *r)
+{
+    size_t body_size, size_digits;
+    u_char *body_size_buf, *body_tag_buf;
+
+    body_size = (r->body) ? nxt_buf_used_size(r->body) : 0;
+    size_digits = (!body_size) ? 1 : log10(body_size) + 1;
+    body_size_buf = nxt_mp_zalloc(r->mem_pool, size_digits + 1);
+    body_tag_buf = nxt_mp_zalloc(r->mem_pool, sizeof(NXT_OTEL_BODY_SIZE_TAG) + 1);
+    if (!body_size_buf || !body_tag_buf) {
+        return;
+    }
```

Comment:
Ugh, I'll come back to this...

However 

```c
sizeof(NXT_OTEL_BODY_SIZE_TAG)
```

Includes the `\0`

But I think it would be safer to do

```c
strlen(NXT_OTEL_BODY_SIZE_TAG) + 1
```

For example if the #define ever got changed into say a `static const char *` where sizeof() would give you the size of a pointer.

Reply:
Good point!

Reply:
Also no need to wrap the ternary conditions in brackets...